File: MergeCheck.cpp
#include "stdafx.h"
#include <iostream>
#include <fstream>
#include <sstream>
#include <map>
using namespace std;

static int listCount = 0, mergeCount = 0, errors = 0;

map<int,int> hashParse(ifstream& file) {
	map<int,int> h;
	cout << "Elements found in merged list (ML):" << endl;
	int i;
	while (file >> i) {
		cout << "ML_" << mergeCount << ": " << i << endl;
		mergeCount++;
		h[i]+=1;
	}
	file.close();
	return h;
} // harshParse

bool compareMerged(ifstream& file,map<int,int>& hash) {
	bool b = true;
	int i;
	while(file >> i) {
		listCount++;
		if (hash.find(i) == hash.end()) {
			b = false;
			cout << " Element " << i << " from sorted lists not found in merged list." << endl;
			errors++;
		}
		else if (hash[i] <= 0) {
			b = false;
			cout << " Copy of Element " << i << " missing in merged list." << endl;
			errors++;
		}
		else hash[i]--;
	}
	file.close();
	return b;
} // compareMerged

int main(int argc, char* argv[])
{

	if (argc < 2) {
		cout << "Run as follows:" << endl;
		cout << " MergeCheck <concatList> <mergedList>" << endl;
		return 0;
	}

	string args[2] = {argv[1], argv[2]};

	ifstream file1(args[0]);
	ifstream file2(args[1]);

	map<int,int> hash;
	hash = hashParse(file2);

	bool consistent = compareMerged(file1,hash);

	if (consistent) cout << "All elements in " << args[0] << " accounted for in " << args[1] << endl;
	else {
		cout << "Error: inconsistency between sorted lists and merged list." << endl;
	}

	cout << "Total number of elements in sorted lists: " << listCount << endl;
	cout << "Total number of elements in merged list: " << mergeCount << endl;
	cout << "Total number of errors detected: " << errors << endl;

	return 0;
} // main

File: MergeCheck.java
import java.io.*;
import java.util.Hashtable;
import java.util.Scanner;


public class MergeCheck {

	static int listCount = 0, mergeCount = 0, errors = 0;
	
	public static void main (String[] args) throws FileNotFoundException {
		
		if (args.length < 2) {
			System.out.println("Run as follows:");
			System.out.println(" java MergeCheck <concatList> <mergedList>");
			return;
		}
		
		File file1 = new File(args[0]);
		File file2 = new File(args[1]);

		Hashtable<Integer,Integer> hash = hashParse(file2);
		boolean consistent = compareMerged(file1,hash);

		if (consistent) System.out.println("All elements in " + args[0] + " accounted for in " + args[1]);
		else {
			System.out.println("Error: inconsistency between sorted lists and merged list.");
			return;
		}
		
		
		System.out.println("Total number of elements in sorted lists: " + listCount);
		System.out.println("Total number of elements in merged list: " + mergeCount);
		System.out.println("Total number of errors detected: " + errors);
	} // main
	
	public static boolean compareMerged(File file, Hashtable<Integer,Integer> hash) throws FileNotFoundException {
	
		boolean b = true;
		Scanner s = new Scanner(file);
		int i;
		while (s.hasNextInt()) {
			i = s.nextInt();
			listCount++;
			if (!hash.containsKey(i)) {
				b = false;
				System.out.println(" Element " + i + " from sorted lists not found in merged list.");
				errors++;
			}
			else if (hash.get(i) <= 0) {
				b = false;
				System.out.println(" Copy of Element " + i + " missing in merged list.");
				errors++;
			}
			else {
				int t = hash.get(i);
				t--;
				hash.put(i, t);
			}
		}
		s.close();
		return b;	
	} // compareMerged
	
	public static Hashtable<Integer, Integer> hashParse(File file) throws FileNotFoundException {
		
		Hashtable<Integer,Integer> h = new Hashtable<Integer, Integer>();
		Scanner reader = new Scanner(file);
		int i;
		System.out.println("Files found in merged list (ML):");
		while (reader.hasNextInt()) {
			i = reader.nextInt();
			System.out.println("ML_" + mergeCount + ": " + i);
			mergeCount++;
			if (!h.containsKey(i)) h.put(i, 1);
			else h.put(i, h.get(i) + 1);
		}
		reader.close();
		return h;
		
	} // hashParse
	
} // MergeCheck
